File: bot/src/bot.ts
import { eq } from "drizzle-orm";
import { db, serverStatesTable, botConfigTable } from "@workspace/db";
import https from "node:https";

const TOKEN = process.env.TELEGRAM_BOT_TOKEN!.trim();
const TELEGRAM_API = `https://api.telegram.org/bot${TOKEN}`;
const CHAT_ID = process.env.TELEGRAM_CHAT_ID;
const BOARD_MSG_KEY = "telegram_board_message_id";

function httpsGet(urlStr: string): Promise<any> {
  return new Promise((resolve, reject) => {
    https.get(urlStr, (res) => {
      let data = "";
      res.on("data", (chunk) => (data += chunk));
      res.on("end", () => {
        try {
          const parsed = JSON.parse(data);
          if (res.statusCode && res.statusCode >= 400) {
            reject(new Error(`HTTP ${res.statusCode}: ${parsed.description || data}`));
          } else {
            resolve(parsed);
          }
        } catch {
          reject(new Error(`Status ${res.statusCode}: ${data}`));
        }
      });
    }).on("error", reject);
  });
}

const SEED_SERVERS = [
  { id: "mio",       name: "Mio" },
  { id: "bo",        name: "Bo" },
  { id: "aaryn",     name: "Aaryn" },
  { id: "new-tunes", name: "New Tunes" },
];

const SERVER_ORDER = ["mio", "bo", "new-tunes", "aaryn"];

async function seedServers(): Promise<void> {
  for (const s of SEED_SERVERS) {
    await db
      .insert(serverStatesTable)
      .values({ id: s.id, name: s.name, status: "off", updatedAt: null })
      .onConflictDoNothing();
  }
}

function toSGT(date: Date): string {
  return date.toLocaleTimeString("en-SG", {
    timeZone: "Asia/Singapore",
    hour: "numeric",
    minute: "2-digit",
    hour12: true,
  });
}

function buildBoard(
  rows: { name: string; status: string; updatedAt: Date | null }[]
): string {
  const sorted = [...rows].sort((a, b) => {
    const aKey = a.name.toLowerCase().replace(" ", "-");
    const bKey = b.name.toLowerCase().replace(" ", "-");
    return SERVER_ORDER.indexOf(aKey) - SERVER_ORDER.indexOf(bKey);
  });

  const lines = sorted.map((r) => {
    if (r.status === "on") {
      const since = r.updatedAt ? toSGT(r.updatedAt) : "unknown";
      return `🔴 ${r.name} — In Use (Since ${since})`;
    }
    return `🟢 ${r.name} — Free`;
  });

  return [
    "📡 *Project Starlight — Live Status*",
    ...lines,
    `🕐 Updated: ${toSGT(new Date())}`,
  ].join("\n");
}

function buildKeyboard(): object {
  const buttons = SERVER_ORDER.map((id, i) => {
    const server = SEED_SERVERS.find((s) => s.id === id)!;
    return { text: `${i + 1} · ${server.name}`, callback_data: `toggle:${id}` };
  });
  return { inline_keyboard: [buttons] };
}

async function telegramPost(method: string, body: object): Promise<any> {
  const res = await fetch(`${TELEGRAM_API}/${method}`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
  return res.json();
}

async function getBoardMessageId(): Promise<number | null> {
  const [row] = await db
    .select()
    .from(botConfigTable)
    .where(eq(botConfigTable.key, BOARD_MSG_KEY));
  return row ? parseInt(row.value, 10) : null;
}

async function saveBoardMessageId(id: number): Promise<void> {
  await db
    .insert(botConfigTable)
    .values({ key: BOARD_MSG_KEY, value: String(id) })
    .onConflictDoUpdate({
      target: botConfigTable.key,
      set: { value: String(id) },
    });
}

export async function sendOrUpdateStatusBoard(): Promise<void> {
  if (!process.env.TELEGRAM_BOT_TOKEN || !CHAT_ID) return;

  try {
    const rows = await db.select().from(serverStatesTable);
    const text = buildBoard(rows);
    const reply_markup = buildKeyboard();
    const boardMessageId = await getBoardMessageId();

    if (boardMessageId === null) {
      const sent = await telegramPost("sendMessage", {
        chat_id: CHAT_ID,
        text,
        parse_mode: "Markdown",
        reply_markup,
      });

      if (sent?.ok && sent?.result?.message_id) {
        await saveBoardMessageId(sent.result.message_id);
        await telegramPost("pinChatMessage", {
          chat_id: CHAT_ID,
          message_id: sent.result.message_id,
          disable_notification: true,
        });
      }
    } else {
      const result = await telegramPost("editMessageText", {
        chat_id: CHAT_ID,
        message_id: boardMessageId,
        text,
        parse_mode: "Markdown",
        reply_markup,
      });

      if (!result?.ok) {
        await db.delete(botConfigTable).where(eq(botConfigTable.key, BOARD_MSG_KEY));
        await sendOrUpdateStatusBoard();
      }
    }
  } catch {
    // Non-fatal
  }
}

export async function handleToggle(serverId: string): Promise<void> {
  const [existing] = await db
    .select()
    .from(serverStatesTable)
    .where(eq(serverStatesTable.id, serverId));

  if (!existing) return;

  const newStatus = existing.status === "on" ? "off" : "on";

  await db
    .update(serverStatesTable)
    .set({ status: newStatus, updatedAt: new Date() })
    .where(eq(serverStatesTable.id, serverId));

  await sendOrUpdateStatusBoard();
}

async function parseUpdate(data: any): Promise<void> {
  console.log("Update received:", JSON.stringify(data).slice(0, 300));
  if (data?.callback_query) {
    const cq = data.callback_query;
    await telegramPost("answerCallbackQuery", { callback_query_id: cq.id }).catch((e) =>
      console.error("answerCallbackQuery failed:", e)
    );
    if (cq.data?.startsWith("toggle:")) {
      const serverId = cq.data.slice("toggle:".length);
      console.log(`Toggle for: ${serverId}`);
      handleToggle(serverId).catch((e) => console.error("handleToggle failed:", e));
    }
    return;
  }

  const text: string = data?.message?.text ?? "";
  if (text === "/start" || text === "/status") {
    sendOrUpdateStatusBoard().catch((e) => console.error("sendOrUpdateStatusBoard failed:", e));
  }
}

let offset = 0;

export async function startPolling(): Promise<void> {
  await seedServers().catch((e) => console.error("seedServers failed:", e));
  await sendOrUpdateStatusBoard().catch((e) => console.error("sendOrUpdateStatusBoard failed:", e));

  console.log("Bot started — polling for updates...");

  let pollCount = 0;
  while (true) {
    ++pollCount;
    try {
      const data = await httpsGet(
        `${TELEGRAM_API}/getUpdates?offset=${offset}&timeout=30`
      );

      if (data?.ok && Array.isArray(data.result)) {
        if (data.result.length > 0) {
          console.log(`[${pollCount}] Got ${data.result.length} update(s)`);
          for (const update of data.result) {
            await parseUpdate(update);
            offset = update.update_id + 1;
          }
        }
      } else {
        console.error(`[${pollCount}] getUpdates not ok:`, JSON.stringify(data).slice(0, 200));
      }
    } catch (err) {
      console.error(`[${pollCount}] Polling error:`, err);
      await new Promise((r) => setTimeout(r, 5000));
    }
  }
}

File: bot/bot.py
import os, asyncio, logging, threading
from datetime import datetime, timezone, timedelta
from http.server import HTTPServer, BaseHTTPRequestHandler

import asyncpg
from telegram import Update, InlineKeyboardButton, InlineKeyboardMarkup
from telegram.ext import Application, CommandHandler, CallbackQueryHandler, ContextTypes

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(message)s")
log = logging.getLogger(__name__)

SGT = timezone(timedelta(hours=8))

TOKEN = os.environ["TELEGRAM_BOT_TOKEN"].strip()
CHAT_ID = os.environ["TELEGRAM_CHAT_ID"].strip()
DATABASE_URL = os.environ["DATABASE_URL"].strip()

SERVERS = {
    "mio": "Mio",
    "bo": "Bo",
    "aaryn": "Aaryn",
    "new-tunes": "New Tunes",
}
SERVER_ORDER = ["mio", "bo", "new-tunes", "aaryn"]
BOARD_MSG_KEY = "telegram_board_message_id"


async def get_pool():
    return await asyncpg.create_pool(DATABASE_URL, min_size=1, max_size=3)


async def seed(pool):
    async with pool.acquire() as conn:
        for sid, name in SERVERS.items():
            await conn.execute(
                "INSERT INTO server_states (id, name, status) VALUES ($1, $2, 'off') ON CONFLICT (id) DO NOTHING",
                sid, name,
            )


async def get_board_msg_id(pool):
    async with pool.acquire() as conn:
        row = await conn.fetchrow(
            "SELECT value FROM bot_config WHERE key = $1", BOARD_MSG_KEY
        )
        return int(row["value"]) if row else None


async def save_board_msg_id(pool, msg_id: int):
    async with pool.acquire() as conn:
        await conn.execute(
            "INSERT INTO bot_config (key, value) VALUES ($1, $2) "
            "ON CONFLICT (key) DO UPDATE SET value = $2",
            BOARD_MSG_KEY, str(msg_id),
        )


async def delete_board_msg_id(pool):
    async with pool.acquire() as conn:
        await conn.execute("DELETE FROM bot_config WHERE key = $1", BOARD_MSG_KEY)


async def get_servers(pool):
    async with pool.acquire() as conn:
        return await conn.fetch("SELECT * FROM server_states")


async def toggle_server(pool, server_id: str):
    async with pool.acquire() as conn:
        row = await conn.fetchrow(
            "SELECT status FROM server_states WHERE id = $1", server_id
        )
        if not row:
            return
        new_status = "off" if row["status"] == "on" else "on"
        await conn.execute(
            "UPDATE server_states SET status = $1, updated_at = NOW() WHERE id = $2",
            new_status, server_id,
        )


def to_sgt(dt):
    if dt is None:
        return None
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=timezone.utc)
    return dt.astimezone(SGT).strftime("%I:%M %p").lstrip("0")


def build_board(rows):
    sorted_rows = sorted(rows, key=lambda r: SERVER_ORDER.index(r["id"]) if r["id"] in SERVER_ORDER else 99)
    lines = []
    for r in sorted_rows:
        if r["status"] == "on":
            since = to_sgt(r["updated_at"]) or "unknown"
            lines.append(f"🔴 {r['name']} — In Use (Since {since})")
        else:
            lines.append(f"🟢 {r['name']} — Free")
    now = datetime.now(timezone.utc)
    updated = to_sgt(now)
    return "\n".join([
        "📡 *Project Starlight — Live Status*",
        *lines,
        f"🕐 Updated: {updated}",
    ])


def build_keyboard():
    buttons = [
        InlineKeyboardButton(f"{i + 1} · {SERVERS[sid]}", callback_data=f"toggle:{sid}")
        for i, sid in enumerate(SERVER_ORDER)
    ]
    return InlineKeyboardMarkup([buttons])


async def send_or_update_board(application, pool):
    servers = await get_servers(pool)
    text = build_board(servers)
    markup = build_keyboard()
    board_msg_id = await get_board_msg_id(pool)

    if board_msg_id is None:
        msg = await application.bot.send_message(
            chat_id=CHAT_ID,
            text=text,
            parse_mode="Markdown",
            reply_markup=markup,
        )
        if msg and msg.message_id:
            await save_board_msg_id(pool, msg.message_id)
            await application.bot.pin_chat_message(
                chat_id=CHAT_ID, message_id=msg.message_id, disable_notification=True
            )
    else:
        try:
            await application.bot.edit_message_text(
                chat_id=CHAT_ID,
                message_id=board_msg_id,
                text=text,
                parse_mode="Markdown",
                reply_markup=markup,
            )
        except Exception:
            await delete_board_msg_id(pool)
            await send_or_update_board(application, pool)


async def cmd_start(update: Update, context: ContextTypes.DEFAULT_TYPE):
    if str(update.effective_chat.id) != CHAT_ID:
        return
    await send_or_update_board(context.application, context.bot_data["pool"])


async def callback_toggle(update: Update, context: ContextTypes.DEFAULT_TYPE):
    if str(update.effective_chat.id) != CHAT_ID:
        return
    query = update.callback_query
    await query.answer()
    server_id = query.data.replace("toggle:", "")
    pool = context.bot_data["pool"]
    await toggle_server(pool, server_id)
    await send_or_update_board(context.application, pool)


class HealthHandler(BaseHTTPRequestHandler):
    def _ok(self):
        self.send_response(200)
        self.end_headers()
        self.wfile.write(b"ok")
    do_GET = _ok
    do_HEAD = _ok
    do_POST = _ok
    do_PUT = _ok
    do_DELETE = _ok
    do_OPTIONS = _ok
    do_PATCH = _ok
    def log_message(self, *a):
        pass


def run_health_server():
    port = int(os.environ.get("PORT", "10000"))
    HTTPServer(("0.0.0.0", port), HealthHandler).serve_forever()


async def main():
    pool = await get_pool()
    await seed(pool)

    app = Application.builder().token(TOKEN).build()
    app.bot_data["pool"] = pool

    app.add_handler(CommandHandler("start", cmd_start))
    app.add_handler(CommandHandler("status", cmd_start))
    app.add_handler(CallbackQueryHandler(callback_toggle, pattern="^toggle:"))

    await send_or_update_board(app, pool)
    log.info("Starting polling...")
    await app.run_polling(allowed_updates=["message", "callback_query"])


if __name__ == "__main__":
    threading.Thread(target=run_health_server, daemon=True).start()
    asyncio.run(main())
